sdio/imx6ull/imx6ull-sdio.c

```diff
@@ -0,0 +1,855 @@
+		}
+	}
+	if (platform_setPad(pctl_pad_csi_vsync, 0, 0, 0, 0, 0, 2, 1, 0) < 0) {
+		return -1;
+	}
+	if (platform_setDevClk(pctl_clk_usdhc2, 3) < 0) {
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static int sdhc_reset(uint32_t mode)
+{
+	uint8_t i;
+	uint32_t val;
+
+	if (mode == SDHC_SYSCTL_RESET_ALL) {
+		*(sdio_common.base + reg_sys_ctrl) = SDHC_SYSCTL_CLK_DEFAULT | SDHC_SYSCTL_RESERVED;
+	}
+	*(sdio_common.base + reg_sys_ctrl) |= mode;
+
+	for (i = 0; i < SDHC_RETRIES; ++i) {
+		usleep(10);
+		val = *(sdio_common.base + reg_sys_ctrl) & mode;
+		if (val == 0) {
+			return 0;
+		}
+	}
+
+	printf("libsdio: sdhc block reset failed\n");
+
+	return -1;
+}
+
+
+static int sdhc_cmdWait(uint32_t flags, uint32_t wait_us, uint32_t max_cnt)
+{
+	uint32_t i, val;
+
+	for (i = 0; i < max_cnt; ++i) {
+		val = *(sdio_common.base + reg_int_status);
+
+		if (val & SDHC_CMD_ERROR) {
+			return -1;
+		}
+
+		if ((val & flags) == flags) {
+			return 0;
+		}
+
+		usleep(wait_us);
+	}
+
+	return -1;
+}
+
+
+static int sdio_isr(unsigned int n, void *arg)
+{
+
+	/* clear external interrupt enable only
+	 * i.MX 6ULL documentation, page 4038 */
+	uint32_t val = *(sdio_common.base + reg_int_status);
+
+	if (val & SDIO_IRQ_STS_CARD) {
+		*(sdio_common.base + reg_int_signal_en) &= ~SDIO_IRQ_STS_CARD;
+	}
+
+	return 0;
+}
+
+
+static void *sdio_dmammap(void)
+{
+	void *p;
+
+	size_t sz = (DMA_BUFFER_SIZE + _PAGE_SIZE - 1) & ~(_PAGE_SIZE - 1);
+
+	p = mmap(NULL, sz, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_UNCACHED, NULL, 0);
+	if (p == MAP_FAILED) {
+		return NULL;
+	}
+
+	return p;
+}
+
+
+static addr_t sdio_mphys(void *p, size_t *psz)
+{
+	size_t sz;
+	addr_t pa;
+
+	pa = va2pa(p);
+	sz = _PAGE_SIZE - (pa & (_PAGE_SIZE - 1));
+
+	while (sz < *psz) {
+		addr_t npa = va2pa(p + sz);
+		if (npa != pa + sz) {
+			break;
+		}
+
+		sz += _PAGE_SIZE;
+	}
+
+	if (sz < *psz) {
+		*psz = sz;
+	}
+
+	return pa;
+}
+
+
+static int sdio_allocDMA(void)
+{
+	size_t psize = DMA_BUFFER_SIZE;
+
+	sdio_common.dmaptr = sdio_dmammap();
+	if (sdio_common.dmaptr == NULL) {
+		return -1;
+	}
+
+	sdio_common.dmaphys = sdio_mphys(sdio_common.dmaptr, &psize);
+
+	if (DMA_BUFFER_SIZE != psize) {
+		printf("libsdio: DMA buffer size is not equal to page size (%zu != %zu)\n", DMA_BUFFER_SIZE, psize);
+		munmap(sdio_common.dmaptr, DMA_BUFFER_SIZE);
+		sdio_common.dmaptr = NULL;
+		return -1;
+	}
+
+	if (sdio_common.dmaphys & 3) {
+		printf("libsdio: DMA buffer physical address is not aligned (%x)\n", sdio_common.dmaphys);
+		munmap(sdio_common.dmaptr, DMA_BUFFER_SIZE);
+		sdio_common.dmaptr = NULL;
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static void sdio_eventThread(void *arg)
+{
+	uint8_t i;
+	uint32_t val;
+
+	mutexLock(sdio_common.eventLock);
+	while (1) {
+		condWait(sdio_common.eventCond, sdio_common.eventLock, 0);
+
+		/* check which event fired if any */
+		val = *(sdio_common.base + reg_int_status);
+		if (val == 0) {
+			continue;
+		}
+		for (i = 0; i < sizeof(eventBits); i++) {
```

Comment:
It's fine since, `eventBits` is an array of `uint8_t`, which are 1 byte in size, but you have defined a macro `ARRAY_LENGTH()` so why not use it here?

Reply:
Replaced with `ARRAY_LENGTH` macro.